network/ipfs_impl.go

```diff
@@ -86,24 +88,128 @@ type impl struct {
 }
 
 type streamMessageSender struct {
-	s     network.Stream
-	bsnet *impl
+	to     peer.ID
+	stream network.Stream
+	bsnet  *impl
+	opts   *MessageSenderOpts
+	done   chan struct{}
 }
 
-func (s *streamMessageSender) Close() error {
-	return helpers.FullClose(s.s)
+// Open a stream to the remote peer
+func (s *streamMessageSender) Connect(ctx context.Context) (network.Stream, error) {
+	if s.stream != nil {
+		return s.stream, nil
+	}
+
+	if err := s.bsnet.ConnectTo(ctx, s.to); err != nil {
+		return nil, err
+	}
+
+	// Check if the sender has been closed
+	select {
+	case <-s.done:
+		return nil, nil
+	default:
+	}
+
+	stream, err := s.bsnet.newStreamToPeer(ctx, s.to)
+	if err != nil {
+		return nil, err
+	}
+
+	s.stream = stream
+	return s.stream, nil
 }
 
+// Reset the stream
 func (s *streamMessageSender) Reset() error {
-	return s.s.Reset()
+	if s.stream != nil {
+		err := s.stream.Reset()
+		s.stream = nil
+		return err
+	}
+	return nil
 }
 
-func (s *streamMessageSender) SendMsg(ctx context.Context, msg bsmsg.BitSwapMessage) error {
-	return s.bsnet.msgToStream(ctx, s.s, msg)
+// Close the stream
+func (s *streamMessageSender) Close() error {
+	close(s.done)
+	return helpers.FullClose(s.stream)
 }
 
+// Indicates whether the peer supports HAVE / DONT_HAVE messages
 func (s *streamMessageSender) SupportsHave() bool {
-	return s.bsnet.SupportsHave(s.s.Protocol())
+	return s.bsnet.SupportsHave(s.stream.Protocol())
+}
+
+// Send a message to the peer, attempting multiple times
+func (s *streamMessageSender) SendMsg(ctx context.Context, msg bsmsg.BitSwapMessage) error {
+	return s.multiAttempt(ctx, func(fnctx context.Context) error {
+		return s.send(fnctx, msg)
+	})
+}
+
+// Perform a function with multiple attempts, and a timeout
+func (s *streamMessageSender) multiAttempt(ctx context.Context, fn func(context.Context) error) error {
+	// Try to call the function repeatedly
+	var err error
+	for i := 0; i < s.opts.MaxRetries; i++ {
+		deadline := time.Now().Add(s.opts.SendTimeout)
+		sndctx, cancel := context.WithDeadline(ctx, deadline)
+
+		if err = fn(sndctx); err == nil {
+			cancel()
+			// Attempt was successful
+			return nil
+		}
+		cancel()
+
```

Comment:
Let's check the error: `errors.Is(err, multistream.ErrNotSupported` (https://github.com/multiformats/go-multistream/blob/bd94c78329f9d3cd394f09e53af6138dd5b4db89/client.go#L11).

If the remote side doesn't support the protocol, we can give up without even trying.